plugins/clickhouse-schema-management/main.go

```diff
@@ -0,0 +1,338 @@
+		versionMap[version] = int(versionNumber - 1)
+	}
+	return nil
+}
+
+// Get Theia version based on the environment variable
+func getTheiaVerstionNumber() (int, error) {
+	theiaVersion := getEnv("THEIA_VERSION")
+	if len(theiaVersion) == 0 {
+		return 0, fmt.Errorf("unable to load environment variables, THEIA_VERSION must be defined")
+	}
+	theiaVersionNumber, err := getVersionNumber(theiaVersion)
+	if err != nil {
+		return theiaVersionNumber, fmt.Errorf("error when getting theia version number for %s: %v", theiaVersion, err)
+	}
+	return theiaVersionNumber, nil
+}
+
+// From v0.3, get data version based on version table
+// For v0.1 and v0.2, determine version based on tables in database
+func getDataVersionNumber(clickhouseMigrate migrate.Migrate) (int, error) {
+	var version int
+	// Get data version for version before v0.3
+	versionStr, err := getDataVersionBasedOnTables()
+	if err != nil {
+		return 0, fmt.Errorf("error when getting data version based on tables: %v", err)
+	}
+	if versionStr != "" {
+		version, err = getVersionNumber(versionStr)
+		if err != nil {
+			return version, fmt.Errorf("error when getting version number for %s: %v", versionStr, err)
+		}
+		if version > 0 {
+			// Set data version for golang-migrate tool
+			err = clickhouseMigrate.Force(version)
+			if err != nil {
+				return 0, fmt.Errorf("error when collarating data version: %v", err)
+			}
+		}
+		return version, nil
+	}
+	// Get data version for version after v0.3
+	uintVersion, _, err := clickhouseMigrate.Version()
+	if err != nil && err != migrate.ErrNilVersion {
+		return version, fmt.Errorf("error when getting migration version: %v", err)
+	}
+	// No data schema created before
+	if err == migrate.ErrNilVersion {
+		return -1, nil
+	}
+	version = int(uintVersion)
+	return version, nil
+}
+
+func getVersionNumber(version string) (int, error) {
+	versionNumber, ok := versionMap[version]
+	if !ok {
+		var err error
+		// In case data schema does not change between version A and B (assuming B is the later version)
+		// we will not have file xxx_A.up.sql and xxx_A.down.sql
+		// In migration, we can treat version A the same as version C,
+		// which is the first version after A having data schema changes comparing its next version.
+		versionNumber, err = roundUpVersion(version)
+		if err != nil {
+			return versionNumber, fmt.Errorf("error when rounding up version: %v", err)
+		}
+	}
+	return versionNumber, nil
+}
+
+func roundUpVersion(version string) (int, error) {
+	var versionNumber int
+	for key, value := range versionMap {
+		less, err := versionLessThan(key, version)
+		if err != nil {
+			return versionNumber, fmt.Errorf("error when comparing version %s and %s: %v", key, version, err)
+		}
+		if less && (versionNumber < value+1) {
+			versionNumber = value + 1
+		}
+	}
+	return versionNumber, nil
+}
+
+// Return true if version a is earlier than version b.
+func versionLessThan(a, b string) (bool, error) {
```

Comment:
Does this work also with a version number like 0.3.1?

(if I understand correctly this is the Theia version number, not the DB schema version number)

Reply:
Yes, this function splits the Theia version 0.3.1 to (0, 3, 1) and compares the number in the tuple from left to right. In this logic, 0.3.1 is later than 0.3.0 but earlier than 0.4.0.

Reply:
My bad - I misread the code in my previous review!